### PR TITLE
feat[SSAO]: Screen Space Ambient Occlusion

### PR DESCRIPTION
**Pull request by MacXxs**
Implemented:
* Screen Space Ambient Occlusion
* Configurable Range check + bias
* Smooth Radius cutoff
* Gaussian blurred SSAO

**Toggle SSAO with F6**

Example:

|No SSAO|SSAO|
|----|----|
|NoSSAO|SSAO|

### CODE REVIEW

**Review by AlexisDieu — DISMISSED** [review]

File: Source/Shaders/deferred_lighting_fragment.glsl [hunk 1]
```diff
@@ -378,6 +382,7 @@ void main()
             environmentBRDF, numLevels_IBL) * cubemap_intensity;
 
         vec3 color = ambient + Lo + emissiveMat.rgb;
+        color = useSSAO == 1 ? color*ssaoFactor : color;
```
Comment (PaablooCH) [inline Source/Shaders/deferred_lighting_fragment.glsl:385]: if the texture gSSAO is only used here, you can obtain it here only if it's needed
```suggestion
        color = useSSAO == 1 ? color*texture(gSSAO, TexCoord).r : color;
```

File: Source/Shaders/ssao_fragment.glsl [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+                    // 16 // 112 (column 3)
+};
+
+layout(std140, binding = 5) uniform Kernel 
+{
+	vec3 kernelSamples[KERNEL_SIZE];
+	vec3 randomTangents[TANGENT_ROWS][TANGENT_COLS];
+};
+
+mat3 createTangentSpace(const vec3 normal, const vec3 randTangent)
```
Comment (PaablooCH) [inline Source/Shaders/ssao_fragment.glsl:39]: inside common/functions/ibl_functions is another CreateTangentSpace look if you can use it

File: Source/DataModels/PostProcess/SSAO.h [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+		float4 randomTangents[RANDOM_TANGENTS_ROWS][RANDOM_TANGENTS_COLS];
+	};
+
+	GLuint uboKernel = 0;
+	GLuint ssaoFrameBuffer = 0;
+	GLuint gSsao = 0;
+	GLuint ssaoBlurFrameBuffer[2] = { 0, 0 };
+	GLuint gSsaoBlured[2] = { 0, 0 };
+	GLuint positionTexture= 0;
+	GLuint normalTexture = 0;
```
Comment (PaablooCH) [inline Source/DataModels/PostProcess/SSAO.h:50]: move to the constructor

File: Source/Shaders/ssao_fragment.glsl (hunk 2 above)
Comment (MacXxs) [reply Source/Shaders/ssao_fragment.glsl:39]: Nope can't use that one, for ssao is calculated a little bit different 😞

File: Source/Shaders/deferred_lighting_fragment.glsl (hunk 1 above)
Comment (MacXxs) [reply Source/Shaders/deferred_lighting_fragment.glsl]: done

File: Source/DataModels/PostProcess/SSAO.h (hunk 3 above)
Comment (MacXxs) [reply Source/DataModels/PostProcess/SSAO.h]: done

**Review by PaablooCH — DISMISSED** [review]

**Review by AlexisDieu — APPROVED** [review]

**Review by PaablooCH — APPROVED** [review]

**Review by AlexisDieu — APPROVED** [review]

**Comment by AuxPort66** [comment]
Physics is already checking #683, as soon as is merged game will work again and we can merge this too

**Review by AuxPort66 — APPROVED** [review]

**Pull request merged**